### PR TITLE
JSON output PoC

### PR DESCRIPTION
**Pull request by davidhrbac**
@facelessuser  thanks for pyspelling.

We use it to automate tests upon the tree of files. Since the standard output is hard to parse by SW I have created this PoC of JSON output.

Standard output:
```
Misspelled words:
<htmlcontent> ../misal-mirror/m.liturgie.cz/misal/03pust/05_01.htm: html>body>p
--------------------------------------------------------------------------------
Helkiášovu
Joakim
Joakimovou
Joakimově
Joakimu
Judovci
kananejské
lentiškem
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../misal-mirror/m.liturgie.cz/misal/03pust/05_01.htm: html>body>h4
--------------------------------------------------------------------------------
Ez
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../misal-mirror/m.liturgie.cz/misal/03pust/05_01.htm: html>body>p
--------------------------------------------------------------------------------
lidsku
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../misal-mirror/m.liturgie.cz/misal/03pust/04_03.htm: html>body>h4
--------------------------------------------------------------------------------
Iz
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../misal-mirror/m.liturgie.cz/misal/03pust/04_03.htm: html>body>div>p
--------------------------------------------------------------------------------
slitováváš
--------------------------------------------------------------------------------
```

JSON output:
```
{
    "../misal-mirror/m.liturgie.cz/misal/03pust/05_01.htm: html>body>p": [
        "Helkiášovu",
        "Joakim",
        "Joakimovou",
        "Joakimově",
        "Joakimu",
        "Judovci",
        "kananejské",
        "lentiškem",
        "lidsku"
    ],
    "../misal-mirror/m.liturgie.cz/misal/03pust/05_01.htm: html>body>h4": [
        "Ez"
    ],
    "../misal-mirror/m.liturgie.cz/misal/03pust/04_03.htm: html>body>h4": [
        "Iz"
    ],
    "../misal-mirror/m.liturgie.cz/misal/03pust/04_03.htm: html>body>div>p": [
        "slitováváš"
    ]
}
```
I guess it would be nice to have context only as a filename and values as unique list.

Are you willing to incorporate JSON output to pyspelling?

### CODE REVIEW

**Comment by facelessuser** [comment]
I don't think I'm against a JSON output. I think we just need to make sure encoding and such makes sense in the JSON and the output makes sense for multiple tasks etc.

I don't remember how we encoded the output and such, so I'll have to peek into that.

Does it dump to a file? Or just a just dumps a JSON encoded text buffer?

**Comment by davidhrbac** [comment]
It just dumps JSON to output the same way as your output. I don't expect issues with encoding, at least UTF-8 is OK as you can see in PR.

As to JSON keys for HTML reference I can use something like this:

```diff
--- a/pyspelling/__main__.py
+++ b/pyspelling/__main__.py
@@ -1,5 +1,6 @@
 """Main."""
 import json
+import re
 import sys
 import argparse
 from pyspelling import spellcheck, __version__
@@ -58,6 +59,7 @@ def run(config, **kwargs):
     count = 0
 
     json_data = {}
+    m = re.compile("(.*?):\s")
 
     for results in spellcheck(
         config,
@@ -76,10 +78,11 @@ def run(config, **kwargs):
         elif results.words:
             fail = True
             if tojson:
-                if results.context in json_data:
-                    json_data[results.context].extend(results.words)
+                g = m.search(results.context)
+                if g.group(1) in json_data:
+                    json_data[g.group(1)].extend(results.words)
                 else:
-                    json_data[results.context] = results.words
+                    json_data[g.group(1)] = results.words
 
             else:
                 print('Misspelled words:\n<%s> %s' % (results.category, results.context))
```

The story behind this PR is that we want to use JSON output and iterate over keys. Then we take values and search for them and encapsulate within `<mark></mark>` tag so we can visualise it in HTML file.

**Comment by facelessuser** [comment]
Cool, when I get a chance I'll dig a little deeper.

**Comment by davidhrbac** [comment]
Yes, you are right there might be issues with encoding.

```
ERROR: ../misal-mirror/m.liturgie.cz/misal/04triduum/03sobota_x.htm -- 'utf-8' codec can't decode byte 0xf8 in position 575: invalid start byte
{
    "../misal-mirror/m.liturgie.cz/misal/03pust/00_06.htm": [
        "Iz",
        "Ez",
        "Lk",
        "Lévi",
        "slitováváš",
        "Mt"
    ],
```

**Comment by facelessuser** [comment]
> Yes, you are right there might be issues with encoding.

Yeah, that doesn't surprise me. I suspected as much.

**Comment by facelessuser** [comment]
I need to double-check what format all the text is in when we get to the logging part. I think some of it may be raw bytes (some of the debug stuff), but some might be in Unicode which we could ensure is in UTF-8. Whether everything encodes to something sane for every kind of encoding 🤷🏻.

**Comment by facelessuser** [comment]
Okay, the "Results" structure looks like it may return everything as Unicode strings, so we could coerce it all to UTF8 if we wanted.

```console
➜  pyspelling git:(master) ✗ python3 -m pyspelling -n markdown
Results(words=[], context='README.md: html>body>p>a>img[alt]', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p>a>img[alt]', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p>a>img[alt]', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p>a>img[alt]', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p>a>img[alt]', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p>a>img[alt]', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p>img[alt]', category='url-free', error=None)
Results(words=[], context='README.md: html>body>h1', category='url-free', error=None)
Results(words=[], context='README.md: html>body>h2', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
Results(words=['Badd', 'wordz'], context='README.md: html>body>p', category='url-free', error=None)
Misspelled words:
<url-free> README.md: html>body>p
--------------------------------------------------------------------------------
Badd
wordz
--------------------------------------------------------------------------------

Results(words=[], context='README.md: html>body>h2', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
Results(words=[], context='README.md: html>body>h2', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
Results(words=[], context='README.md: html>body>p', category='url-free', error=None)
!!!Spelling check failed!!!
```

Any issues you are seeing with encoding, like below, may be related to not configuring encoding properly for your source input:

```js
ERROR: ../misal-mirror/m.liturgie.cz/misal/04triduum/03sobota_x.htm -- 'utf-8' codec can't decode byte 0xf8 in position 575: invalid start byte
{
    "../misal-mirror/m.liturgie.cz/misal/03pust/00_06.htm": [
        "Iz",
        "Ez",
        "Lk",
        "Lévi",
        "slitováváš",
        "Mt"
    ],
```

**Comment by facelessuser** [comment]
I'll pull this branch locally over the weekend and run some tests on it.

**Comment by facelessuser** [comment]
@davidhrbac  Looking at this pull more, would it be beneficial to also include the error messages in the JSON? Seems that using JSON will hide all the error messages.

**Comment by facelessuser** [comment]
Conversation and work on this have apparently stalled. I'll either have to look into this myself or just close this.

@gir-bot add S: needs-decision

**Comment by facelessuser** [comment]
Closing this for now as there appears to be no interest in pushing this PR forward. It is possible we may decide to do this ourselves in the future, but for now, this PR appears to be stale.

**Comment by davidhrbac** [comment]
@facelessuser sorry for not replying. I'm not maintaining project using pyspeling in CI/CD any more.

**Comment by facelessuser** [comment]
That's fine. There hasn't really been anyone else clambering for this feature, so I won't worry about it right now. If interest picks up again, we may consider it in the future.

Thanks for at least confirming this is no longer required.
